packages/js/components/src/date-time-picker-control/date-time-picker-control.tsx

```diff
@@ -174,72 +174,118 @@ export const DateTimePickerControl: React.FC< DateTimePickerControlProps > = ( {
 	// - [useDebounce hook loses function calls if the dependency changes](https://github.com/WordPress/gutenberg/issues/35505)
 	// - [useMemoOne and useCallbackOne](https://github.com/alexreardon/use-memo-one)
 
-	const onChangePropFunctionRef = useRef<
+	const onChangeRef = useRef<
 		DateTimePickerControlOnChangeHandler | undefined
 	>();
 	useLayoutEffect( () => {
-		onChangePropFunctionRef.current = onChange;
+		onChangeRef.current = onChange;
 	}, [ onChange ] );
 
-	function inputStringChangeHandlerFunction(
-		newInputString: string,
-		fireOnChange: boolean
-	) {
-		if ( ! isMounted.current ) return;
+	function setInputStringWithMoment( dateTime: Moment ) {
+		const newInputString = dateTime.isValid()
+			? formatMoment( dateTime )
+			: dateTime.creationData().input?.toString() || '';
+
+		setInputString( newInputString );
+	}
+
+	function callOnChange( dateTime: Moment ) {
+		const valueToSend = dateTime.isValid()
+			? formatMomentIso( dateTime )
+			: dateTime.creationData().input?.toString() || '';
+
+		if ( typeof onChangeRef.current === 'function' ) {
+			onChangeRef.current( valueToSend, dateTime.isValid() );
+		}
+	}
 
-		let newDateTime = parseMoment( newInputString );
+	function updateState(
```

Comment:
This one function in particular appears to be a generic catch all for updating all things state.  I think it might be more readable to break this apart into multiple function calls.  For example, it's a lot of overhead to go from `updateStateWithNewCurrentDateProp` to `updateStateWithNewDateTime` to `updateState` and then trace backwards and forwards again to see which args were used.

Inside the `useEffect` instead of:
```
		updateStateWithNewCurrentDateProp(
			currentDate || '',
			! isInitialUpdate.current
		);
```

It might be easier to understand:

```
const formattedDateTime = formatDateTime( currentDate );
setInputString( formattedDateTime );
```

Reply:
💯 I reworked things to reduce the nesting of functions and improve the readability of what's actually going on in the code.